### PR TITLE
CP-24744: Add VUSB plug/unplug

### PR DESCRIPTION
**Pull request by taoyongd**
To support usb passthrough feature, will use qemu to
    add/del usb devices to vm.

    1. Add vusbs to vm metadata.
    2. Add VUSB VUSB_DB module.
    3. Add VUSB_plug VUSB_unplug to atom operation.

### CODE REVIEW

**Comment by taoyongd** [comment]
As this PR denpends on the xcp-idl PR https://github.com/xapi-project/xcp-idl/pull/173

File: lib/xenops_server_simulator.ml [hunk 1]
```diff
@@ -256,6 +258,20 @@ let vgpu_state vm vgpu () =
 			| [] -> unplugged_vgpu
 			| _ -> assert false (* at most one *)
 
+let vusb_state vm vusb () =
+	if not (DB.exists vm)
+	then unplugged_vusb
+	else
+		let d = DB.read_exn vm in
+		let this_one x = x.Vusb.id = vusb.Vusb.id in
+		match List.filter this_one d.Domain.vusbs with
```
Comment (lindig) [inline lib/xenops_server_simulator.ml]: Maybe `List.find` is more appropriate.

File: xc/device.ml [hunk 2]
```diff
@@ -1392,6 +1392,82 @@ module Vkbd = struct
...
+    debug "Vusb plugged: vusb device plugged ";
+    let get_bus v =
+      if String.startswith "1" v then
+         "usb-bus.0"
+      else begin
+          (* Here plug usb controller according to the usb version*)
+          let usb_controller_driver = "usb-ehci" in
+          let driver_id = "ehci" in
+          vusb_controller_plug ~xs ~domid ~driver:usb_controller_driver ~driver_id;
+          "ehci.0"
```
Comment (lindig) [inline xc/device.ml]: To avoid any problems when `driver_id` changes:
```
`Printf.sprintf "%s.0" driver_id
```

File: xc/device.ml [hunk 3]
```diff
@@ -1392,6 +1392,82 @@ module Vkbd = struct
...
+    | _ ->()
+
+  let vusb_unplug ~xs ~domid ~id =
+    debug "Vusb unplugged: vusb device unplugged";
+    let is_running = Qemu.is_running ~xs domid in
+    match is_running with
+    | true  -> (
+      let open Qmp in
+      let qmp_cmd = Command (None, Device_del (id)) in
+      debug "qmp unplug usb device";
```
Comment (lindig) [inline xc/device.ml]: Include more information about the device - otherwise the log line is useless

Comment (taoyongd) [reply xc/device.ml]: yes,  here the debug is useless

File: xc/device.ml (hunk 2 above)
Comment (taoyongd) [reply xc/device.ml]: updated. Thanks

File: xc/device.ml (hunk 3 above)
Comment (taoyongd) [reply xc/device.ml]: updated. Thanks

**Comment by taoyongd** [comment]
Hi @jonludlam  @robhoes , could you please also help to review this PR? Thanks a lot.

File: xc/device.ml [hunk 4]
```diff
@@ -1392,6 +1392,81 @@ module Vkbd = struct
...
+    )
+    | _ ->()
+
+  let vusb_plug ~xs ~domid ~id ~hostbus ~hostport ~version =
+    debug "Vusb plugged: vusb device plugged ";
+    let get_bus v =
+      if String.startswith "1" v then
+         "usb-bus.0"
+      else begin
+          (* Here plug usb controller according to the usb version*)
```
Comment (jonludlam) [inline xc/device.ml]: What exactly is the version?

Comment (jonludlam) [reply xc/device.ml]: In fact, it'd be good to know what the range of possible values are for each of the VUSB fields

Comment (jonludlam) [reply xc/device.ml]: OK, answering my own question here. It's the string representation of the bcdUSB field. See http://www.beyondlogic.org/usbnutshell/usb5.shtml for example. It'll take the values: "1.10", "2.00", "2.50", "3.00", "3.10", and maybe more.

**Comment by jonludlam** [comment]
OK, I think this looks fine.

File: xc/device.ml (hunk 4 above)
Comment (taoyongd) [reply xc/device.ml]: yes :)

**Pull request merged**